### ISSUE TITLE
How can I return the RefMut wrapper of a hashmap value in dashmap?

### ISSUE THREAD

**Issue by malongshuai** [xacrimon/dashmap#254]
I Like dashmap features, but sometimes i want to return a mutable reference value in the dashmap(eg. return mut ref of a collection element, which the collection is the value of dashmap),  i don't know how to resolve the lifetime errors.

```rust
use dashmap::{mapref::one::MappedRefMut, DashMap};
use std::{sync::Arc, collections::HashMap};

#[derive(Debug, Clone)]
struct User {
    name: String,
    age: u32,
}

type SomeUsers = HashMap<String, User>;

#[derive(Clone, Default)]
pub struct UserManager {
    active_users: Arc<DashMap<String, SomeUsers>>,
}

impl UserManager {
    fn user_mut(&self, k1: &str, k2: &str) -> MappedRefMut<'_, String, SomeUsers, Option<&mut User>> {
        let n = self.active_users.get_mut(k1).unwrap();
        n.map(|x| &mut x.get_mut(k2))
    }
}
```

compile error like this:
```
error: lifetime may not live long enough
  --> src/new_main.rs:20:19
   |
20 |         n.map(|x| &mut x.get_mut(k2))
   |                -- ^^^^^^^^^^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
   |                ||
   |                |return type of closure is &mut Option<&'2 mut User>
   |                has type `&'1 mut std::collections::HashMap<String, User>`
   |
   = note: requirement occurs because of a mutable reference to `Option<&mut User>`
   = note: mutable references are invariant over their type parameter
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance

error[E0515]: cannot return reference to temporary value
  --> src/new_main.rs:20:19
   |
20 |         n.map(|x| &mut x.get_mut(k2))
   |                   ^^^^^-------------
   |                   |    |
   |                   |    temporary value created here
   |                   returns a reference to data owned by the current function
```

How can I get the mutable reference of the User in HashMap in DashMap?

**Comment by hniksic** [comment]
Your attempt was very close, `RefMut::map()` is almost the right tool for the job. The problem was not in the lifetime, but in that you wanted to return an `Option`, and `RefMut::map()` expects the function to be infallible. If you were willing to unwrap the inner `get_mut()`, this would work just fine:

```rust
// note generic type is User, not &mut User
pub fn user_mut(&self, k1: &str, k2: &str) -> MappedRefMut<'_, String, SomeUsers, User>{
    let n = self.active_users.get_mut(k1).unwrap();
    n.map(|x| x.get_mut(k2).unwrap())
}
```

Trying to make map fallible by changing `User` to `Option<&mut User>` doesn't work because that would require the `Option<&mut User>` to be stored somewhere in the map, which it's not (`User` is stored). Instead, you can use `try_map()`, which exists for that purpose. Your return value will then be `Option<MappedRefMut<...>>`, which is probably a better API in the first place:

```rust
pub fn user_mut(&self, k1: &str, k2: &str) -> Option<MappedRefMut<'_, String, SomeUsers, User>>{
    let n = self.active_users.get_mut(k1).unwrap();
    n.try_map(|x| x.get_mut(k2)).ok()
}
```

In summary, this issue can probably be closed, `DashMap` already supports this use case nicely through `RefMut::try_map()`.
